File: SudokuSolver_Java_No_Streams/src/main/sudokusolver/javanostreams/SudokuNumber.java
package sudokusolver.javanostreams;

/*
 * The only collection that should be used for SudokuNumber is EnumSet. Sometimes this will lead to more complicated
 * code to avoid using a List or Set.
 */
public enum SudokuNumber {
    ONE, TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE;

    @Override
    public String toString() {
        return Integer.toString(ordinal() + 1);
    }

    public static SudokuNumber valueOf(char ch) {
        return switch (ch) {
            case '1' -> ONE;
            case '2' -> TWO;
            case '3' -> THREE;
            case '4' -> FOUR;
            case '5' -> FIVE;
            case '6' -> SIX;
            case '7' -> SEVEN;
            case '8' -> EIGHT;
            case '9' -> NINE;
            default -> throw new IllegalArgumentException("ch is '" + ch + "', must be between '1' and '9'.");
        };
    }
}

File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/SudokuNumber.kt
package sudokusolver.kotlin

/*
 * The only collection that should be used for SudokuNumber is EnumSet. Sometimes this will lead to more complicated
 * code to avoid using a List or Set.
 */
enum class SudokuNumber {
    ONE, TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE;

    override fun toString(): String = "${ordinal + 1}"
}

fun sudokuNumber(ch: Char): SudokuNumber =
    when (ch) {
        '1' -> SudokuNumber.ONE
        '2' -> SudokuNumber.TWO
        '3' -> SudokuNumber.THREE
        '4' -> SudokuNumber.FOUR
        '5' -> SudokuNumber.FIVE
        '6' -> SudokuNumber.SIX
        '7' -> SudokuNumber.SEVEN
        '8' -> SudokuNumber.EIGHT
        '9' -> SudokuNumber.NINE
        else -> throw IllegalArgumentException("ch is '$ch', must be between '1' and '9'.")
    }

fun parseOptionalBoard(board: String): Board<SudokuNumber?> {
    require(board.length == UNIT_SIZE_SQUARED) { "board length is ${board.length}, must be $UNIT_SIZE_SQUARED." }
    return Board(board.chunked(UNIT_SIZE) { row ->
        row.map { cell -> cell.takeUnless { it == '0' }?.let { sudokuNumber(it) } }
    })
}

fun parseBoard(board: String): Board<SudokuNumber> {
    require(board.length == UNIT_SIZE_SQUARED) { "board length is ${board.length}, must be $UNIT_SIZE_SQUARED." }
    return Board(board.chunked(UNIT_SIZE) { row -> row.map { sudokuNumber(it) } })
}